src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+		assert( sourceInterp == PrimitiveVariable::Vertex || sourceInterp == PrimitiveVariable::Varying );
+
+		const MeshPrimitive *sourceMesh = IECore::runTimeCast< const MeshPrimitive >( sourcePrim );
+		const std::vector<int> &sourceVertexIds = sourceMesh->vertexIds()->readable();
+
+		for( size_t j = 0; j < numIndices; j++ )
+		{
+			*(destIndices++) = translateIndex( sourceVertexIds[ j ] );
+		}
+	}
+}
+
+} // namespace
+
+
+void PrimitiveAlgo::transformPrimitive(
+	IECoreScene::Primitive &primitive, Imath::M44f matrix,
+	const IECore::Canceller *canceller
+)
+{
+	if( matrix == Imath::M44f() )
+	{
+		// Early out for identity matrix
+		return;
+	}
+
+	Imath::M44f normalMatrix = normalTransform( matrix );
+
+	for( const auto &[name, var] : primitive.variables )
+	{
+		Canceller::check( canceller );
+		V3fVectorData *vecVar = IECore::runTimeCast<V3fVectorData>( var.data.get() );
+		V3fData *vecConstVar = IECore::runTimeCast<V3fData>( var.data.get() );
+		if( !vecVar && !vecConstVar )
+		{
+			continue;
+		}
+
+		GeometricData::Interpretation interp = vecVar ? vecVar->getInterpretation() : vecConstVar->getInterpretation();
+		if( !(
+			interp == GeometricData::Interpretation::Point ||
+			interp == GeometricData::Interpretation::Vector ||
+			interp == GeometricData::Interpretation::Normal
+		) )
+		{
+			continue;
+		}
+
+		if( vecVar )
+		{
+			for( Imath::V3f &i : vecVar->writable() )
+			{
+				Canceller::check( canceller );
+				transformPrimVarValue( i, i, matrix, normalMatrix, interp );
+			}
+		}
+		else
+		{
+			// Fairly weird corner case, but technically Constant primvars could need transforming too
+			transformPrimVarValue( vecConstVar->writable(), vecConstVar->writable(), matrix, normalMatrix, interp );
+		}
+	}
+}
+
+IECoreScene::PrimitivePtr PrimitiveAlgo::mergePrimitives(
+	const std::vector< std::pair< const IECoreScene::Primitive*, Imath::M44f > > &primitives,
+	const IECore::Canceller *canceller
+)
+{
+	IECoreScene::TypeId resultTypeId = (IECoreScene::TypeId)IECore::InvalidTypeId;
+
+	// Mesh specific globals
+	std::string resultMeshInterpolation;
+
+	// Curve specific globals
+	const CubicBasisf invalidBasis( Imath::M44f( 0.0f ), 0 );
+	CubicBasisf resultCurvesBasis( invalidBasis );
+	bool resultCurvesPeriodic( false );
+
+	// Data we need to store for each primvar we output
+	struct PrimVarInfo
+	{
+		PrimVarInfo( IECore::TypeId t, GeometricData::Interpretation interpretation, int numMeshes )
+			: invalid( false ), needsVerts( false ), needsFaces( false ), needsCurveKnots( false ),
+			typeId( t ), interpret( interpretation ), interpretInvalid( false ), indexed( false ),
+			numData( numMeshes, 0 )
+		{
+		}
+		// Primitive variables that are invalid should be reported once when we first realize it's invalid,
+		// and subsequently ignored.
+		bool invalid;
+
+		// These flags determine what interpolation is necessary for the result
+		bool needsVerts;
+		bool needsFaces;
+		bool needsCurveKnots;
```

Comment:
As a new reader of this (huge) function, one tricky thing is understanding what data is used and where, since many variables carry through to the very end. These ones are an example of something that is only needed temporarily (to figure out the interpolation) but then live till the end of the function. If considering refactoring, I think a focus on limiting each thing to its useful life would be helpful.

Reply:
This has been totally reworked.